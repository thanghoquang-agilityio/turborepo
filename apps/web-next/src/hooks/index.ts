export * from './useBreadcrumb'
export * from './useCart'
export * from './useDebounce'
export * from './useOrderSummary'
export * from './usePagination'
