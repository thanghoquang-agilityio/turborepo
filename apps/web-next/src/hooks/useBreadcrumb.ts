import { usePathname } from 'next/navigation'

import { BreadcrumbItem } from '@/components'
import { PRODUCT_LIST_REGEX, ROUTER } from '@/constants'

export const useBreadcrumb = () => {
  const pathname = usePathname()

  const newPath = pathname?.split('/').filter((path) => path)
  const pathLength = newPath.length

  const parentPage = pathname.match(PRODUCT_LIST_REGEX)
    ? {
        name: 'Products',
        link: ROUTER.PRODUCTS,
      }
    : ({} as BreadcrumbItem)

  const itemsBreadcrumb: BreadcrumbItem[] = [parentPage]

  if (pathLength > 1 && pathname.match(PRODUCT_LIST_REGEX))
    itemsBreadcrumb.push({
      name: 'Detail Products',
      link: '#',
    })

  return itemsBreadcrumb
}
