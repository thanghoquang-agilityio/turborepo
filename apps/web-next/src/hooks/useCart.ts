'use client'

import { useCallback, useState } from 'react'

// Actions
import {
  createOrUpdateCartItem,
  deleteCartItem,
  updateCartItem,
} from '@/app/actions'
// constants
import { CART_MESSAGES } from '@/constants'
// Toasts
import { useToast } from '@/context/toast'
import { TOAST_TYPE } from '@/hocs/withToast'
// Types
import { CartItemPayload } from '@/types'

export const useCart = () => {
  const openToast = useToast()
  const [isPending, setIsPending] = useState(false)

  const handleQuantityChange = useCallback(
    async (documentId: string, cartItem: CartItemPayload) => {
      const response = await updateCartItem(documentId, cartItem)

      const { FAILED, SUCCESS } = CART_MESSAGES.UPDATE
      const message = response ? SUCCESS : FAILED

      openToast({
        type: response ? TOAST_TYPE.SUCCESS : TOAST_TYPE.ERROR,
        message,
      })

      return response
    },
    [openToast]
  )

  const handleRemove = useCallback(
    async (documentId: string) => {
      setIsPending(true)

      const isSuccess = await deleteCartItem(documentId)

      const { FAILED, SUCCESS } = CART_MESSAGES.REMOVE
      const message = isSuccess ? SUCCESS : FAILED

      setIsPending(false)

      openToast({
        type: isSuccess ? TOAST_TYPE.SUCCESS : TOAST_TYPE.ERROR,
        message,
      })

      return isSuccess
    },
    [openToast]
  )

  const handleCreateOrUpdate = useCallback(
    async (cartItem: CartItemPayload) => {
      setIsPending(true)
      const response = await createOrUpdateCartItem(cartItem)
      setIsPending(false)

      const { FAILED, SUCCESS, WARNING } = CART_MESSAGES.ADD
      const message =
        response === null ? FAILED : response?.isOverStock ? WARNING : SUCCESS

      openToast({
        type:
          response === null
            ? TOAST_TYPE.ERROR
            : response?.isOverStock
              ? TOAST_TYPE.WARNING
              : TOAST_TYPE.SUCCESS,
        message,
      })

      return response
    },
    [openToast]
  )
  return {
    isPending,
    handleQuantityChange,
    handleRemove,
    handleCreateOrUpdate,
  }
}
