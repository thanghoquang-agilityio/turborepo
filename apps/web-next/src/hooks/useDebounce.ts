import { useRef } from 'react'

export const useDebounce = <T>(callback: (params: T) => void, delay = 1000) => {
  const timerRef = useRef<NodeJS.Timeout | null>(null)

  return function (params: T) {
    // Clear the previous timer
    if (timerRef.current) {
      clearTimeout(timerRef.current)
    }

    timerRef.current = setTimeout(() => {
      callback(params)
    }, delay)
  }
}
