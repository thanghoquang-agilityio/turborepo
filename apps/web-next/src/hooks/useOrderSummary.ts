'use client'

import { useOptimistic } from 'react'

import { CartItemResponse } from '@/types'
import { calculateAmount } from '@/utils'

export const useOrderSummary = (cartItems: CartItemResponse[]) => {
  const [optimisticCartItems, addOptimisticCartItems] = useOptimistic(
    cartItems,
    (state: CartItemResponse[], newCartItem: CartItemResponse) => {
      const existingItemIndex = state.findIndex(
        (item) => item.id === newCartItem.id
      )

      if (existingItemIndex !== -1) {
        const updatedState = [...state]
        updatedState[existingItemIndex] = {
          ...updatedState[existingItemIndex],
          quantity: newCartItem.quantity,
        }

        return updatedState
      } else {
        return [...state, newCartItem]
      }
    }
  )

  const { totalAmount, totalItems } = optimisticCartItems.reduce(
    (acc, cartItem) => {
      const { quantity = 0, productVariantId } = cartItem
      const { tax = 0 } = productVariantId || {}
      
      // Handle product relationship in new API structure
      // These values could come from:
      // 1. Extended cart item with product data: cartItem.productDiscount, cartItem.productPrice
      // 2. Product data on variant: productVariantId?.product?.discount, productVariantId?.product?.price
      // 3. Separate product data passed to hook
      const discount = 0 // cartItem.productDiscount || productVariantId?.product?.discount || 0
      const price = 0 // cartItem.productPrice || productVariantId?.product?.price || 0

      const amount = calculateAmount(tax, discount, price)

      return {
        totalAmount: acc.totalAmount + amount.price * quantity,
        totalItems: acc.totalItems + quantity,
      }
    },
    { totalAmount: 0, totalItems: 0 }
  )

  return {
    totalAmount,
    totalItems,
    addOptimisticCartItems,
  }
}
