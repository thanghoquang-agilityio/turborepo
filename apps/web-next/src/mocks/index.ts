export * from './breadcrumb'
export * from './cart'
export * from './category'
export * from './image'
export * from './product'
export * from './radio-group'
export * from './user'
