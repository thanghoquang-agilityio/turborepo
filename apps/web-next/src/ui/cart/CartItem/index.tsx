'use client'

import { memo, useCallback, useMemo, useState } from 'react'

import { Card as CardNextUI, Spinner } from '@repo/next-ui'

// Components
import { Button, CustomImage, Text } from '@/components'
// Constants
import { API_IMAGE_URL, USER_ID_DEFAULT, borderStyle } from '@/constants'
// Hooks
import { useCart, useDebounce } from '@/hooks'
import { CloseIcon, MinusIcon, PlusIcon } from '@/icons'
// Styles
import '@/styles/card.css'
// Types
import type { CartItemResponse } from '@/types'
// Utils
import { calculateAmount, formatAmountCurrency } from '@/utils'

interface CartItemProps {
  documentId: string
  variantId: string
  image?: string
  name: string
  color: string
  size: string
  quantityDefault: number
  stock: number
  price: number
  tax: number
  discount: number
  additionalClass?: string
  addOptimisticCartItems: (action: CartItemResponse) => void
}

const CartItem = memo(
  ({
    documentId,
    variantId,
    image,
    name,
    color,
    size,
    quantityDefault,
    stock,
    price,
    tax,
    discount,
    additionalClass = '',
    addOptimisticCartItems,
  }: CartItemProps) => {
    const [quantity, setQuantity] = useState(quantityDefault)
    const [oldQuantity, setOldQuantity] = useState(quantityDefault)

    const cartItem = useMemo(() => {
      return {
        id: 0, // Not used for updates
        documentId,
        productVariantId: Number(variantId),
        userId: Number(USER_ID_DEFAULT),
        quantity,
        size,
        createdAt: '',
        updatedAt: '',
        publishedAt: '',
      }
    }, [documentId, quantity, size, variantId])

    const { isPending, handleQuantityChange, handleRemove } = useCart()
    const debounceQuantityChange = useDebounce(async (value: number) => {
      let res
      if (!value) {
        addOptimisticCartItems({
          ...cartItem,
          productVariantId: undefined,
          quantity: 0,
        })

        res = await handleRemove(documentId)
      } else if (value !== oldQuantity) {
        addOptimisticCartItems({
          ...cartItem,
          productVariantId: undefined,
          quantity: value,
        })

        res = await handleQuantityChange(documentId, {
          productVariantId: cartItem.productVariantId,
          userId: cartItem.userId || 0,
          size: cartItem.size,
          quantity: value,
        })
      }

      if (!res) setQuantity(oldQuantity)
      else setOldQuantity(value)
    })

    // Handle change quantity
    const onChange = useCallback(
      async (operation: 'increment' | 'decrement') => {
        const quantityChange =
          operation === 'increment' ? quantity + 1 : quantity - 1

        setQuantity(quantityChange)
        debounceQuantityChange(quantityChange)
      },
      [debounceQuantityChange, quantity]
    )

    const handleDecrement = useCallback(async () => {
      onChange('decrement')
    }, [onChange])

    const handleIncrement = useCallback(async () => {
      onChange('increment')
    }, [onChange])

    // Handle remove item
    const onRemove = useCallback(async () => {
      addOptimisticCartItems({
        ...cartItem,
        productVariantId: undefined,
        quantity: 0,
      })

      await handleRemove(documentId)
    }, [addOptimisticCartItems, cartItem, handleRemove, documentId])

    const amount = useMemo(
      () => calculateAmount(tax, discount, price),
      [discount, price, tax]
    )

    const altImage = name.split('')[0]
    const subTotal = amount.price * quantity
    const formatSubTotal = useMemo(
      () => formatAmountCurrency(subTotal.toString()),
      [subTotal]
    )

    return (
      <CardNextUI
        className={`card relative flex max-w-[305px] flex-col items-center overflow-hidden rounded-large border-medium border-background-300 pb-5 pt-3 ${additionalClass}`}
      >
        {isPending && (
          <>
            <div className='absolute inset-0 z-50 flex items-center justify-center bg-foreground-100 opacity-40' />
            <Spinner
              size='lg'
              classNames={{
                circle1:
                  'border-[5px] border-b-primary-100 border-l-primary-100',
                circle2:
                  'border-[5px] border-b-primary-100 border-l-primary-100',
              }}
              className='z-60 absolute inset-0 m-auto'
            />
          </>
        )}
        <Button
          variant='ghost'
          additionalClass='absolute top-0 right-0 text-primary-300'
          data-testid='button-close'
          onClick={onRemove}
        >
          <CloseIcon customClass='w-[12px] h-[12px]' />
        </Button>
        <Text
          variant='primary'
          className='max-w-[250px] text-xl text-primary-400'
        >
          {name}
        </Text>
        <Text
          variant='subTitle'
          size='sm'
          className='max-w-[250px] capitalize text-secondary-500'
        >
          {color} / {size}
        </Text>
        <CustomImage
          src={`${API_IMAGE_URL}${image}`}
          alt={altImage}
          className='card-image mb-5 mt-4 h-[303px] w-[305px]'
          width={305}
          height={303}
        />

        <div className='flex h-[40px] w-[150px] justify-between rounded-[30px] border-small border-background-300 bg-background-100 p-2'>
          <Button
            variant='ghost'
            additionalClass='h-full text-primary-100'
            isDisabled={quantity === 0}
            data-testid='button-minus'
            onClick={handleDecrement}
          >
            <MinusIcon customClass='w-[12px] h-[12px]' />
          </Button>
          <Text className='text-[20px] font-bold leading-[22px] text-primary-300'>
            {quantity}
          </Text>
          <Button
            variant='ghost'
            additionalClass='h-full text-primary-100'
            isDisabled={quantity === stock}
            data-testid='button-plus'
            onClick={handleIncrement}
          >
            <PlusIcon customClass='w-[12px] h-[12px]' />
          </Button>
        </div>

        <div
          className={`relative mt-10 flex w-full justify-between px-4 ${borderStyle} before:top-[-20px]`}
        >
          <Text variant='title' size='lg' className='max-w-[100px]'>
            Subtotal
          </Text>
          <Text variant='title' size='lg' className='max-w-[100px]'>
            {formatSubTotal}
          </Text>
        </div>
      </CardNextUI>
    )
  }
)

CartItem.displayName = 'CartItem'
export default CartItem
