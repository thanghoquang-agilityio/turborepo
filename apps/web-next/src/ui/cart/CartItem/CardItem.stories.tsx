import type { Meta, StoryObj } from '@storybook/react'

import { MOCK_CART_ITEM } from '@/mocks'

import CartItem from '../CartItem'

const meta = {
  title: 'Components/CartItem',
  tags: ['autodocs'],
  component: CartItem,
  parameters: {
    layout: 'centered',
  },
} satisfies Meta<typeof CartItem>

export default meta
type Story = StoryObj<typeof CartItem>

export const Default: Story = {
  args: MOCK_CART_ITEM,
}
