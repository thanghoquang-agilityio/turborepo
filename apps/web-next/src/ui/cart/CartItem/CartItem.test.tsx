import { fireEvent, render, waitFor } from '@testing-library/react'

import { useCart } from '@/hooks'
import { MOCK_CART_ITEM } from '@/mocks'

import CartItem from '.'

// Mock hooks
jest.mock('@/hooks', () => ({
  useCart: jest.fn(),
  useDebounce: jest.fn((fn) => fn), // Assuming that useDebounce is called with a function.
}))

const mockHandleQuantityChange = jest.fn()
const mockHandleRemove = jest.fn()

// Explicitly set the mock implementation for useCart
;(useCart as jest.Mock).mockReturnValue({
  handleQuantityChange: mockHandleQuantityChange,
  handleRemove: mockHandleRemove,
})

// Mock utils functions
jest.mock('@/utils', () => ({
  formatAmountCurrency: jest.fn((value) => `$${value}`),
  calculateAmount: jest.fn(() => ({
    price: 100, // Mocked return value of calculateAmount
  })),
}))

describe('CartItem', () => {
  const renderComponent = () =>
    render(<CartItem {...MOCK_CART_ITEM} addOptimisticCartItems={() => {}} />)

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should match snapshot', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('should increment quantity and call debounceQuantityChange', async () => {
    const { getByTestId } = renderComponent()
    const incrementButton = getByTestId('button-plus')

    fireEvent.click(incrementButton)

    await waitFor(() => {
      expect(mockHandleQuantityChange).toHaveBeenCalledWith(
        MOCK_CART_ITEM.id,
        expect.objectContaining({
          quantity: MOCK_CART_ITEM.quantityDefault + 1,
        })
      )
    })
  })

  it('should decrement quantity and call debounceQuantityChange', async () => {
    const { getByTestId } = renderComponent()
    const decrementButton = getByTestId('button-minus')

    fireEvent.click(decrementButton)

    await waitFor(() => {
      expect(mockHandleQuantityChange).toHaveBeenCalledWith(
        MOCK_CART_ITEM.id,
        expect.objectContaining({
          quantity: MOCK_CART_ITEM.quantityDefault - 1,
        })
      )
    })
  })

  it('should handle removal of item and show spinner', async () => {
    const { getByTestId } = renderComponent()
    const removeButton = getByTestId('button-close')

    fireEvent.click(removeButton)

    await waitFor(() => {
      // Check that the remove handler was called
      expect(mockHandleRemove).toHaveBeenCalledWith(MOCK_CART_ITEM.id)
    })
  })
})
