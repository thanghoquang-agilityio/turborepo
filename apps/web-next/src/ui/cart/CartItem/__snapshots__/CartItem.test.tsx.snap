// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem should match snapshot 1`] = `
<div>
  <div
    class="h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium transition-transform-background motion-reduce:transition-none card relative flex items-center flex-col max-w-[305px] pt-3 pb-5 border-medium border-background-300 rounded-large overflow-hidden"
    tabindex="-1"
  >
    <button
      aria-label="Button"
      class="z-0 group inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 h-10 text-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover gap-2.5 font-semibold bg-transparent border-none min-w-3 min-h-5 rounded-full absolute top-0 right-0 text-primary-300"
      data-testid="button-close"
      type="button"
    >
      <svg
        class="w-[12px] h-[12px]"
        fill="none"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M376.6 84.5c11.3-13.6 9.5-33.8-4.1-45.1s-33.8-9.5-45.1 4.1L192 206 56.6 43.5C45.3 29.9 25.1 28.1 11.5 39.4S-3.9 70.9 7.4 84.5L150.3 256 7.4 427.5c-11.3 13.6-9.5 33.8 4.1 45.1s33.8 9.5 45.1-4.1L192 306 327.4 468.5c11.3 13.6 31.5 15.4 45.1 4.1s15.4-31.5 4.1-45.1L233.7 256 376.6 84.5z"
          fill="currentColor"
        />
      </svg>
    </button>
    <p
      class="text-primary-300 font-semibold  no-underline text-ellipsis whitespace-nowrap overflow-hidden text-xl text-primary-400 max-w-[250px]"
    >
      Embrace Sideboard
    </p>
    <p
      class="text-foreground-300 font-normal text-sm no-underline text-ellipsis whitespace-nowrap overflow-hidden max-w-[250px] text-secondary-500 capitalize"
    >
      wheat
       / 
      small
    </p>
    <img
      alt="E"
      class="card-image w-[305px] h-[303px] mt-4 mb-5"
      data-nimg="1"
      decoding="async"
      fetchpriority="high"
      height="303"
      src="http://localhost/_next/image?url=https%3A%2F%2Fnext-js-p6iu.onrender.comundefined&w=640&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fnext-js-p6iu.onrender.comundefined&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fnext-js-p6iu.onrender.comundefined&w=640&q=75 2x"
      style="color: transparent; object-fit: cover;"
      width="305"
    />
    <div
      class="flex w-[150px] justify-between p-2 h-[40px] rounded-[30px] bg-background-100 border-small border-background-300"
    >
      <button
        aria-label="Button"
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 text-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover gap-2.5 font-semibold bg-transparent border-none min-w-3 min-h-5 rounded-full h-full text-primary-100"
        data-testid="button-minus"
        type="button"
      >
        <svg
          class="w-[12px] h-[12px]"
          fill="none"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 256c0 17.7-14.3 32-32 32L48 288c-17.7 0-32-14.3-32-32s14.3-32 32-32l352 0c17.7 0 32 14.3 32 32z"
            fill="currentColor"
          />
        </svg>
      </button>
      <p
        class="  no-underline text-ellipsis whitespace-nowrap overflow-hidden text-primary-300 font-bold text-[20px] leading-[22px]"
      >
        2
      </p>
      <button
        aria-label="Button"
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 text-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover gap-2.5 font-semibold bg-transparent border-none min-w-3 min-h-5 rounded-full h-full text-primary-100"
        data-testid="button-plus"
        type="button"
      >
        <svg
          class="w-[12px] h-[12px]"
          fill="none"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 144L48 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l144 0 0 144c0 17.7 14.3 32 32 32s32-14.3 32-32l0-144 144 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-144 0 0-144z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="relative flex justify-between w-full mt-10 px-4 before:content-[''] before:absolute before:w-full before:border-t-small before:border-background-300 before:top-[-20px]"
    >
      <p
        class="text-primary-200 dark:text-primary-300 font-semibold text-lg no-underline text-ellipsis whitespace-nowrap overflow-hidden max-w-[100px]"
      >
        Subtotal
      </p>
      <p
        class="text-primary-200 dark:text-primary-300 font-semibold text-lg no-underline text-ellipsis whitespace-nowrap overflow-hidden max-w-[100px]"
      >
        $200
      </p>
    </div>
  </div>
</div>
`;
